### DISCUSSION TITLE
Parse symbols inside EchoNode

### DISCUSSION THREAD

**Discussion by khalyomede** [Stillat/blade-parser#12] [Ideas]
My need is the following: I am (still) working on this Laravel package to pull languages strings from a blade file, and I want to support not only `@lang()` and `@choice()` (working fine thanks to this package), but also `{{ trans() }}`, `{{ __() }}` and `{{ trans_choice() }}`.

For the moment, this is how the following blade content will be parsed

```blade
@extends("layout")

@section("content")
  {{ trans("List of books") }}
@endsection
```

```php
Stillat\BladeParser\Nodes\EchoNode^ {#2652 // /home/ubuntu/app/src/Commands/Translate.php:96
  +index: 8
  +id: "N6818fb73cf6a-677e-4a0b-a2b7-4a2758b31465"
  +content: "{{ trans('List of books') }}"
```

What I would have needed is the `childNodes` (?) to contain a list of all parsed nodes.

Here is my proposal:

1. We get all the raw content between `{{` and `}}`
2. We call nikic/php-parser and get the AST from it
3. We adapt it to fit blade-parser AST

In my experiment, here is how `trans('List of books');` renders

```php
use PhpParser\NodeDumper;
use PhpParser\ParserFactory;

require __DIR__ . "/vendor/autoload.php";

$parser = (new ParserFactory)->create(ParserFactory::PREFER_PHP7);
$ast = $parser->parse("<?php trans('List of books');");
$dumper = new NodeDumper();

echo $dumper->dump($ast);
```

```
array(
    0: Stmt_Expression(
        expr: Expr_FuncCall(
            name: Name(
                parts: array(
                    0: trans
                )
            )
            args: array(
                0: Arg(
                    name: null
                    value: Scalar_String(
                        value: List of books
                    )
                    byRef: false
                    unpack: false
                )
            )
        )
    )
)
```

Willing to make a pull request if the idea appeals to you (might take some round trips between you and me untill it is polished as you wish).

**Comment by JohnathonKoster** [comment]
I'm not entirely opposed to the idea, but will need to do some thinking before pulling that in as a dependency (even though its fantastic). The `childNodes` array will be populated with any nodes that appear between directive pairs (or Blade component open/close tags) when `resolveStructures` is called.

The nodes are macroable, however, and this is something you could add to your current project:

```php
<?php

use Illuminate\Support\Str;
use PhpParser\ParserFactory;
use Stillat\BladeParser\Document\Document;
use Stillat\BladeParser\Nodes\EchoNode;

$template = <<<'BLADE'
@extends("layout")

@section("content")
{{ trans("List of books") }}
@endsection
BLADE;

EchoNode::macro('toAst', function () {
    $php = '<?php '.Str::finish($this->innerContent, ';').' ?>';

    return (new ParserFactory)->create(ParserFactory::PREFER_PHP7)->parse($php);
});

Document::fromText($template)->getEchoes()->each(function (EchoNode $node) {
    $ast = $node->toAst();

    dd($ast);
});
```
